### PR TITLE
Add feature policy for wake lock

### PR DESCRIPTION
**Pull request by Honry**
https://github.com/w3c/wake-lock/issues/111

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/20069)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/295595435)

**Comment by Honry** [comment]
@marcoscaceres, @andrey-logvinov, PTAL.

File: wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+<script src=/feature-policy/resources/featurepolicy.js></script>
+<script>
+  'use strict';
+
+  const relative_path = '/feature-policy/resources/feature-policy-wakelock.html';
+  const base_src = '/feature-policy/resources/redirect-on-load.html#';
+  const same_origin_src = base_src + relative_path;
+  const cross_origin_src = base_src + 'https://{{domains[www]}}:{{ports[https][0]}}' +
+      relative_path;
+  const header = 'Feature-Policy allow="wake-lock"';
```
[the remark this answers is not in the document]
Comment (andrey-logvinov) [reply wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html]: I see now, it is done by passing "wake-lock" as `feature_name` to `test_feature_availability`.

**Review by andrey-logvinov — APPROVED** [review]

File: feature-policy/resources/feature-policy-wakelock.html [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<script>
+'use strict';
+
+Promise.resolve().then(() => navigator.getWakeLock("screen")).then(wakelock => {
```
Comment (marcoscaceres) [inline feature-policy/resources/feature-policy-wakelock.html]: Nit: do we really need "Promise.resolve().then("?

Comment (marcoscaceres) [inline feature-policy/resources/feature-policy-wakelock.html]: I'd suggest changing this to an async function, with try/catch. Can we do anything useful with that error?

File: feature-policy/resources/feature-policy-wakelock.html [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<script>
```
Comment (marcoscaceres) [inline feature-policy/resources/feature-policy-wakelock.html]: Can you check the naming for this file? If it's supposed to be used in an iframe, you might want to call it ".sub.html". See other tests that do this. That will help distinguish this file from actual test.

File: wake-lock/wakelock-disabled-by-feature-policy.https.sub.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+<script src=/feature-policy/resources/featurepolicy.js></script>
+<script>
+  'use strict';
+
+  const same_origin_src = '/feature-policy/resources/feature-policy-wakelock.html';
+  const cross_origin_src = 'https://{{domains[www]}}:{{ports[https][0]}}' +
+      same_origin_src;
+  const header = 'Feature-Policy header {"wake-lock" : []}';
+
+  promise_test(() => {
```
Comment (marcoscaceres) [inline wake-lock/wakelock-disabled-by-feature-policy.https.sub.html]: switch to async, and make this a try/catch, with the error being `unreached_assertion()`.

File: wake-lock/wakelock-disabled-by-feature-policy.https.sub.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script src=/feature-policy/resources/featurepolicy.js></script>
+<script>
+  'use strict';
+
+  const same_origin_src = '/feature-policy/resources/feature-policy-wakelock.html';
+  const cross_origin_src = 'https://{{domains[www]}}:{{ports[https][0]}}' +
+      same_origin_src;
+  const header = 'Feature-Policy header {"wake-lock" : []}';
```
Comment (marcoscaceres) [inline wake-lock/wakelock-disabled-by-feature-policy.https.sub.html]: Better to just inline this.

File: wake-lock/wakelock-enabled-by-feature-policy.https.sub.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+<script>
+  'use strict';
+
+  const same_origin_src = '/feature-policy/resources/feature-policy-wakelock.html';
+  const cross_origin_src = 'https://{{domains[www]}}:{{ports[https][0]}}' +
+      same_origin_src;
+  const header = 'Feature-Policy header {"wake-lock" : ["*"]}';
+
+  promise_test(
+      () => navigator.getWakeLock("screen"),
```
Comment (marcoscaceres) [inline wake-lock/wakelock-enabled-by-feature-policy.https.sub.html]: Nit: whitespace on this test makes it confusing to read.

File: wake-lock/wakelock-enabled-by-feature-policy.https.sub.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+<script src=/resources/testharnessreport.js></script>
+<script src=/feature-policy/resources/featurepolicy.js></script>
+
+<script>
+  'use strict';
+
+  const same_origin_src = '/feature-policy/resources/feature-policy-wakelock.html';
+  const cross_origin_src = 'https://{{domains[www]}}:{{ports[https][0]}}' +
+      same_origin_src;
+  const header = 'Feature-Policy header {"wake-lock" : ["*"]}';
```
Comment (marcoscaceres) [inline wake-lock/wakelock-enabled-by-feature-policy.https.sub.html]: As above.

**Review by marcoscaceres — COMMENTED** [review]
Couple of comments... will review more closely once questions from @andrey-logvinov are addressed.

File: feature-policy/resources/feature-policy-wakelock.html (hunk 3 above)
Comment (Honry) [reply feature-policy/resources/feature-policy-wakelock.html]: [Tests involves multiple origins](http://web-platform-tests.org/writing-tests/server-features.html#tests-involving-multiple-origins) should use ".sub" postfix, this is just a resource file, we can leave it as is.

File: feature-policy/resources/feature-policy-wakelock.html (hunk 2 above)
Comment (Honry) [reply feature-policy/resources/feature-policy-wakelock.html]: I have one concern here, what is the error should be, "SecurityError"? But the spec does not explicitly state.

File: wake-lock/wakelock-disabled-by-feature-policy.https.sub.html (hunk 4 above)
Comment (Honry) [reply wake-lock/wakelock-disabled-by-feature-policy.https.sub.html]: I would prefer `promise_rejects` if we know what the error should be.

File: wake-lock/wakelock-disabled-by-feature-policy.https.sub.html (hunk 5 above)
Comment (Honry) [reply wake-lock/wakelock-disabled-by-feature-policy.https.sub.html]: Sure.

File: wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<body>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script src=/feature-policy/resources/featurepolicy.js></script>
+<script>
+  'use strict';
+
+  const relative_path = '/feature-policy/resources/feature-policy-wakelock.html';
+  const base_src = '/feature-policy/resources/redirect-on-load.html#';
```
[the remark this answers is not in the document]
Comment (Honry) [reply wake-lock/wakelock-enabled-by-feature-policy-attribute-redirect-on-load.https.sub.html]: I think there is no difference, I just leverage the test framework from https://github.com/w3c/web-platform-tests/blob/master/feature-policy/README.md

File: feature-policy/resources/feature-policy-wakelock.html (hunk 2 above)
Comment (marcoscaceres) [reply feature-policy/resources/feature-policy-wakelock.html]: Good catch! Should definitely add to the spec.

File: wake-lock/wakelock-disabled-by-feature-policy.https.sub.html (hunk 4 above)
Comment (marcoscaceres) [reply wake-lock/wakelock-disabled-by-feature-policy.https.sub.html]: Yeah, good point. I agree.

**Comment by Honry** [comment]
@marcoscaceres, @andrey-logvinov, thanks for your review, I've addressed your comments, pls. check again.

File: feature-policy/resources/feature-policy-wakelock.html (hunk 3 above)
Comment (marcoscaceres) [reply feature-policy/resources/feature-policy-wakelock.html]: Thanks for clarifying. I think then I screwed up somewhere else 😊

**Review by marcoscaceres — APPROVED** [review]
Did some style fixes.

**Comment by Honry** [comment]
Wow... it looks beautiful now, really appreciate! @marcoscaceres.

**Comment by marcoscaceres** [comment]
Keep up the great work, @Honry! Looking forward to the next batch of tests from you.

**Comment by marcoscaceres** [comment]
Btw, I'm just using [Prettier](https://github.com/prettier/prettier) via Visual Studio Code to format the JS. I just copy the JS into a new document, run prettier over the JS, and then paste is back in. It's a little bit annoying to do manually, but totally worth it for consistency.

**Pull request merged**